No ssl-context via `SslContext` in Aleph 0.7.0 and later

When trying to upgrade one of our services from Aleph 0.6.4 to 0.8.0 I saw this strange exception in the logs:

```
[...]
Caused by: clojure.lang.ExceptionInfo: Some desired HTTP versions are not part of ALPN config.
{:desired-http-versions [:http1], :alpn-protocols []}
 at aleph.http.common$assert_consistent_alpn_config_BANG_.invokeStatic (common.clj:139)
    aleph.http.common$assert_consistent_alpn_config_BANG_.invoke (common.clj:130)
    aleph.http.common$ensure_consistent_alpn_config.invokeStatic (common.clj:182)
    aleph.http.common$ensure_consistent_alpn_config.invoke (common.clj:143)
    aleph.http.client$client_ssl_context.invokeStatic (client.clj:698)
    aleph.http.client$client_ssl_context.invoke (client.clj:690)
    aleph.http.client$http_connection.invokeStatic (client.clj:799)
    aleph.http.client$http_connection.invoke (client.clj:752)
    aleph.http$create_connection.invokeStatic (http.clj:104)
    aleph.http$create_connection.invoke (http.clj:97)
    aleph.http$connection_pool$fn__21882.invoke (http.clj:239)
    aleph.flow$instrumented_pool$reify__12626.generate (flow.clj:47)
    io.aleph.dirigiste.Pool.addObject (Pool.java:273)
    io.aleph.dirigiste.Pool.acquire (Pool.java:466)
    aleph.flow$acquire$fn__12631.invoke (flow.clj:74)
    aleph.flow$acquire.invokeStatic (flow.clj:73)
    aleph.flow$acquire.invoke (flow.clj:68)
    aleph.http$eval21907$request__21911$fn__21915$fn__21916.invoke (http.clj:377)
    aleph.http$eval21907$request__21911$fn__21915.invoke (http.clj:371)
    aleph.http$eval21907$request__21911.invoke (http.clj:370)
[...]
```

When trying to pin this down, I could confirm that this already happens in Aleph 0.7.0 (even for the alpha versions) and it only happens when the pool is configured with an `io.netty.handler.ssl.SslContext` instance. I.e.

```clojure
(def ssl-context (.build (io.netty.handler.ssl.SslContextBuilder/forClient)))

(def pool (http/connection-pool {:connection-options {:ssl-context ssl-context}}))

(http/get "https://example.com" {:pool pool}) =>
#<Deferred@346c5293: Error printing return value (ExceptionInfo) at aleph.http.common/assert-consistent-alpn-config! (common.clj:139).
Some desired HTTP versions are not part of ALPN config.
```

Note that the exception happens even before the deferred is dereferenced. Also note that I just use a plain `io.netty.handler.ssl.JdkSslClientContext` instance here without any special SSL config in this example, since the specific config doesn't seem to matter here.

If I use a plain map to initialize the `ssl-context`, e.g. say

```clojure
(def ssl-context {:trust-store (io/as-file client-ca)})
```

(whereas `io/as-file` is `clojure.java.io/as-file` here, and `client-ca` a file path string), this doesn't happen. So that's a viable workaround. Still, I couldn't find any update in the documentation that says `SslClientContext` instances are not supported anymore, so this is at least a documentation "bug".

---

### Addendum

As per the docs, instead of `java.io.File` instances, `java.io.InputStream` instances are also supported as keys for the `:ssl-context` map. But I can't figure out how this is supposed to be used, as I regularly get `java.lang.IllegalArgumentException: Input stream does not contain valid certificates.` exceptions. I.e., preparing a single request just works fine, but preparing them in quick succession may fail:

```clojure
(def ssl-context {:trust-store (io/input-stream client-ca)})

(def pool (http/connection-pool {:connection-options {:ssl-context ssl-context}}))

(http/get "https://example.com" {:pool pool}) => #<Deferred@6fb6283a: :not-delivered>

[(http/get "https://example.com" {:pool pool}) (http/get "https://example.com" {:pool pool})] =>
[#<Deferred@58ac8429: Error printing return value (CertificateException) at io.netty.handler.ssl.PemReader/readCertificates (PemReader.java:114).
Error printing return value (CertificateException) at io.netty.handler.ssl.PemReader/readCertificates (PemReader.java:114).
found no certificates in input stream
            PemReader.java:  114  io.netty.handler.ssl.PemReader/readCertificates
           SslContext.java: 1263  io.netty.handler.ssl.SslContext/toX509Certificates
    SslContextBuilder.java:  276  io.netty.handler.ssl.SslContextBuilder/trustManager
                 netty.clj:  917  aleph.netty/eval23500/add-ssl-trust-manager!
                 netty.clj: 1026  aleph.netty/eval23500/ssl-client-context
                 netty.clj: 1193  aleph.netty/coerce-ssl-context
                 netty.clj: 1179  aleph.netty/coerce-ssl-context
                  core.clj: 2641  clojure.core/partial/fn
                client.clj:  699  aleph.http.client/client-ssl-context
                client.clj:  690  aleph.http.client/client-ssl-context
                client.clj:  799  aleph.http.client/http-connection
                client.clj:  752  aleph.http.client/http-connection
                  http.clj:  104  aleph.http/create-connection
                  http.clj:   97  aleph.http/create-connection
                  http.clj:  239  aleph.http/connection-pool/fn
                  flow.clj:   47  aleph.flow/instrumented-pool/reify
                 Pool.java:  273  io.aleph.dirigiste.Pool/addObject
                 Pool.java:  466  io.aleph.dirigiste.Pool/acquire
                  flow.clj:   74  aleph.flow/acquire/fn
                  flow.clj:   73  aleph.flow/acquire
                  flow.clj:   68  aleph.flow/acquire
                  http.clj:  377  aleph.http/eval30155/request/fn/fn
                  http.clj:  371  aleph.http/eval30155/request/fn
                  http.clj:  370  aleph.http/eval30155/request
                  http.clj:  481  aleph.http/req
                  http.clj:  477  aleph.http/req
                  core.clj: 2642  clojure.core/partial/fn
```

I didn't look into the implementation, but I suspect what's happening here is that when the first thread of the thread pool is initialized, it's exhausting the input-stream instance and this instance is reused during the initialization of a second thread. (At least that's what I hope is happening, since the alternative would be that each thread tries to reread the certificate on each request.)

So the question is, if this doesn't work, why is it even supported? But if this is indeed an issue, it probably should be handled in a separate ticket, since this behavior already applies to Aleph 0.6.4 and therefore can't be considered a regression.

--- a/src/aleph/http.clj
+++ b/src/aleph/http.clj
@@ -208,11 +208,17 @@
            control-period       60000
            middleware           middleware/wrap-request
            max-queue-size       65536}}]
-  (when (and (false? (:keep-alive? connection-options))
-             (pos? (:idle-timeout connection-options 0)))
-    (throw
-      (IllegalArgumentException.
+  (let [{:keys [keep-alive?
+                idle-timeout
+                http-versions
+                force-h2c?]
+         :or {idle-timeout 0}} connection-options]
+    (when (and (false? keep-alive?) (pos? idle-timeout))
+      (throw
+       (IllegalArgumentException.
         ":idle-timeout option is not allowed when :keep-alive? is explicitly disabled")))
+    (when (and force-h2c? (not-any? #{:http2} http-versions))
+      (throw (IllegalArgumentException. "force-h2c? may only be true when HTTP/2 is enabled."))))
 
   (let [log-activity (:log-activity connection-options)
         dns-options' (if-not (and (some? dns-options)

--- a/src/aleph/http/common.clj
+++ b/src/aleph/http/common.clj
@@ -194,7 +194,7 @@
         ssl-context)
       (if (= [:http1] desired-http-versions)
         ssl-context
-        (throw (ex-info "No ALPN supplied, but requested non-HTTP/1 versions that require ALPN."
+        (throw (ex-info "Supplied SslContext with no ALPN config, but requested secure non-HTTP/1 versions that require ALPN."
                         {:desired-http-versions desired-http-versions}))))))
 
 (defn validate-http1-pipeline-transform

--- a/src/aleph/http/common.clj
+++ b/src/aleph/http/common.clj
@@ -158,8 +158,9 @@
    If `ssl-context` is an options map and it does contain an `:application-protocol-config` key, its
    supported protocols are validated to match `desired-http-versions`.
 
-   Otherwise, if `ssl-context` is an `io.netty.handler.ssl.SslContext` instance, its supported
-   protocols are validated to match `desired-http-versions`."
+   Otherwise, if `ssl-context` is an `io.netty.handler.ssl.SslContext` instance, its ALPN config's
+   protocols are validated to match `desired-http-versions`. If it doesn't have an ALPN config,
+   `desired-http-versions` may only contain `:http1`."
   [ssl-context desired-http-versions]
   (when-let [unsupported-http-versions (seq (remove supported-http-versions desired-http-versions))]
     (throw (ex-info "Unsupported desired HTTP versions"
@@ -178,13 +179,23 @@
     nil
 
     :else
-    (do
-      (assert-consistent-alpn-config!
-       (some-> ^SslContext ssl-context
-               (.applicationProtocolNegotiator)
-               (.protocols))
-       desired-http-versions)
-      ssl-context)))
+    ;; NOTE: `SslContext` always has an `ApplicationProtocolNegotiator`, even if
+    ;; `.applicationProtocolConfig` was never invoked in the `SslContextBuilder`. In this case, its
+    ;; `.protocols` will be an empty collection, which we thus treat as if no ALPN config is
+    ;; present.
+    (if-let [protocols (some-> ^SslContext ssl-context
+                               (.applicationProtocolNegotiator)
+                               (.protocols)
+                               (seq))]
+      (do
+        (assert-consistent-alpn-config!
+         protocols
+         desired-http-versions)
+        ssl-context)
+      (if (= [:http1] desired-http-versions)
+        ssl-context
+        (throw (ex-info "No ALPN supplied, but requested non-HTTP/1 versions that require ALPN."
+                        {:desired-http-versions desired-http-versions}))))))
 
 (defn validate-http1-pipeline-transform
   "Checks that :pipeline-transform is not being used with HTTP/2, since Netty

--- a/test/aleph/http_test.clj
+++ b/test/aleph/http_test.clj
@@ -1339,7 +1339,7 @@
     (catch Exception e
       e)))
 
-(deftest test-http-versions-config
+(deftest test-http-versions-server-config
   (testing "ssl-context as options map"
 
     (testing "with different HTTP versions in ALPN config"
@@ -1424,12 +1424,18 @@
                                           (netty/application-protocol-config [:http2 :http1])))})]
         (is (= :started result))))
 
-    (testing "with no ALPN config"
+    (testing "with no ALPN config but desiring HTTP/2"
       (let [result (try-start-server
                     {:http-versions [:http2 :http1]
                      :ssl-context test-ssl/server-ssl-context})]
         (is (instance? ExceptionInfo result))
-        (is (= "Some desired HTTP versions are not part of ALPN config." (ex-message result))))))
+        (is (= "No ALPN supplied, but requested non-HTTP/1 versions that require ALPN." (ex-message result)))))
+
+    (testing "with no ALPN config but desiring only HTTP/1"
+      (let [result (try-start-server
+                    {:http-versions [:http1]
+                     :ssl-context test-ssl/server-ssl-context})]
+        (is (= :started result)))))
 
   (testing "HTTP/2 without ssl-context"
     (let [result (try-start-server

--- a/test/aleph/http_test.clj
+++ b/test/aleph/http_test.clj
@@ -1429,7 +1429,7 @@
                     {:http-versions [:http2 :http1]
                      :ssl-context test-ssl/server-ssl-context})]
         (is (instance? ExceptionInfo result))
-        (is (= "No ALPN supplied, but requested non-HTTP/1 versions that require ALPN." (ex-message result)))))
+        (is (= "Supplied SslContext with no ALPN config, but requested secure non-HTTP/1 versions that require ALPN." (ex-message result)))))
 
     (testing "with no ALPN config but desiring only HTTP/1"
       (let [result (try-start-server
@@ -1583,7 +1583,7 @@
                        {:http-versions [:http2 :http1]
                         :ssl-context test-ssl/client-ssl-context}})]
           (is (instance? ExceptionInfo result))
-          (is (= "No ALPN supplied, but requested non-HTTP/1 versions that require ALPN." (ex-message result)))))
+          (is (= "Supplied SslContext with no ALPN config, but requested secure non-HTTP/1 versions that require ALPN." (ex-message result)))))
 
       (testing "with no ALPN config but desiring only HTTP/1"
         (let [result (try-request-with-pool

--- a/test/aleph/http_test.clj
+++ b/test/aleph/http_test.clj
@@ -1464,6 +1464,154 @@
       (is (instance? IllegalArgumentException result))
       (is (= "use-h2c? may only be true when HTTP/2 is enabled." (ex-message result))))))
 
+(defn ex-root-cause [e]
+  (if-let [c (ex-cause e)]
+    (recur c)
+    e))
+
+(defn try-request-with-pool [options]
+  (try
+    (let [^Pool pool (http/connection-pool options)]
+      (try
+        @(http-get "/" {:pool pool})
+        :success
+        (finally
+          (.shutdown pool))))
+    (catch Exception e
+      ;; Exceptions raised during connection creation are wrapped in multiple levels of
+      ;; `RuntimeException` which we don't care about here.
+      (ex-root-cause e))))
+
+(deftest test-http-versions-client-config
+  (with-http-ssl-servers echo-handler {}
+    (testing "ssl-context as options map"
+
+      (testing "with different HTTP versions in ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2]
+                        :ssl-context (assoc test-ssl/client-ssl-context-opts
+                                            :application-protocol-config
+                                            (netty/application-protocol-config [:http1]))}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "Some desired HTTP versions are not part of ALPN config." (ex-message result)))))
+
+      (testing "with different preference order in ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2 :http1]
+                        :ssl-context (assoc test-ssl/client-ssl-context-opts
+                                            :application-protocol-config
+                                            (netty/application-protocol-config [:http1 :http2]))}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "Desired HTTP version preference order differs from ALPN config." (ex-message result)))))
+
+      (testing "with extra HTTP versions in the ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http1]
+                        :ssl-context (assoc test-ssl/client-ssl-context-opts
+                                            :application-protocol-config
+                                            (netty/application-protocol-config [:http1 :http2]))}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "ALPN config contains more HTTP versions than desired." (ex-message result)))))
+
+      (testing "with matching ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2 :http1]
+                        :ssl-context (assoc test-ssl/client-ssl-context-opts
+                                            :application-protocol-config
+                                            (netty/application-protocol-config [:http2 :http1]))}})]
+          (is (= :success result))))
+
+      (testing "with no ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2 :http1]
+                        :ssl-context test-ssl/client-ssl-context-opts}})]
+          (is (= :success result)))))
+
+    (testing "ssl-context as SslContext instance"
+
+      (testing "with different HTTP versions in ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2]
+                        :ssl-context (netty/ssl-client-context
+                                      (assoc test-ssl/client-ssl-context-opts
+                                             :application-protocol-config
+                                             (netty/application-protocol-config [:http1])))}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "Some desired HTTP versions are not part of ALPN config." (ex-message result)))))
+
+      (testing "with different preference order in ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2 :http1]
+                        :ssl-context (netty/ssl-client-context
+                                      (assoc test-ssl/client-ssl-context-opts
+                                             :application-protocol-config
+                                             (netty/application-protocol-config [:http1 :http2])))}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "Desired HTTP version preference order differs from ALPN config." (ex-message result)))))
+
+      (testing "with extra HTTP versions in the ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http1]
+                        :ssl-context (netty/ssl-client-context
+                                      (assoc test-ssl/client-ssl-context-opts
+                                             :application-protocol-config
+                                             (netty/application-protocol-config [:http1 :http2])))}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "ALPN config contains more HTTP versions than desired." (ex-message result)))))
+
+      (testing "with matching ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2 :http1]
+                        :ssl-context (netty/ssl-client-context
+                                      (assoc test-ssl/client-ssl-context-opts
+                                             :application-protocol-config
+                                             (netty/application-protocol-config [:http2 :http1])))}})]
+          (is (= :success result))))
+
+      (testing "with no ALPN config but desiring HTTP/2"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2 :http1]
+                        :ssl-context test-ssl/client-ssl-context}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "No ALPN supplied, but requested non-HTTP/1 versions that require ALPN." (ex-message result)))))
+
+      (testing "with no ALPN config but desiring only HTTP/1"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http1]
+                        :ssl-context test-ssl/client-ssl-context}})]
+          (is (= :success result)))))
+
+    (testing "HTTP/2 without ssl-context"
+      (let [result (try-request-with-pool
+                    {:connection-options
+                     {:http-versions [:http2]
+                      ;; We only want assert that not providing `:ssl-context` is possible but since
+                      ;; the test server uses a self-signed certificate, we cannot verify it with
+                      ;; the default one. Since all we care about is ALPN here, this is fine.
+                      :insecure? true}})]
+        (is (= :success result)))))
+
+  (with-http2-server echo-handler {:use-h2c? true :ssl-context nil}
+    (with-redefs [*use-tls-requests* false]
+      (testing "h2c with HTTP/2"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:force-h2c? true
+                        :http-versions [:http2]}})]
+          (is (= :success result)))))))
+
+
 (deftest test-in-flight-request-cancellation
   (let [conn-established (promise)
         conn-closed (promise)]

--- a/test/aleph/http_test.clj
+++ b/test/aleph/http_test.clj
@@ -1609,7 +1609,15 @@
                       {:connection-options
                        {:force-h2c? true
                         :http-versions [:http2]}})]
-          (is (= :success result)))))))
+          (is (= :success result))))
+
+      (testing "h2c without HTTP/2"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:force-h2c? true
+                        :http-versions [:http1]}})]
+          (is (instance? IllegalArgumentException result))
+          (is (= "force-h2c? may only be true when HTTP/2 is enabled." (ex-message result))))))))
 
 
 (deftest test-in-flight-request-cancellation

